release-19.2: storage: allow healing ranges whose replicas are all lost

tbg opened the pull request:
The goal is to run the following cli command (as of this commit):

```
./cockroach debug unsafe-heal-missing-range <rangeid> --host=[...]
```

This needs to be run against a node *also running this commit*
(otherwise, it will fail gracefully). The cli command will hit
an endpoint on the node that synthesizes a snapshot for the
range which has only that node as a member, and applies it
to itself. The range is then functional again and should
replicate as indicated by replication factor and constraints,
which should be verified via the range status.

The safest way to apply this fix is the following:

1. Add a new node to the cluster (running this commit) and immediately
decommission it.
2. Run the above command against this fresh node.

This approach has the benefit that the node carrying out the work
is not an essential part of the cluster. This means that unexpected
crashes or corruption would likely be confined to this one node.

If this is deemed too much work, an existing node can be used.

In either case, it is indisposable to have taken a recent backup and
to be ready to respond to unanticipated consequences (for example
by restoring the backup).

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/50387)
<!-- Reviewable:end -->

tbg commented:
The acceptance test passed (and I checked that it did run the new test). I had tested this locally prior as well.

tbg committed:
release-19.2: helper to recover from range with all replicas lost

The goal is to run the following cli command (as of this commit):

```
./cockroach debug unsafe-heal-missing-range <rangeid> --host=[...]
```

This needs to be run against a node *also running this commit*
(otherwise, it will fail gracefully). The cli command will hit
an endpoint on the node that synthesizes a snapshot for the
range which has only that node as a member, and applies it
to itself. The range is then functional again and should
replicate as indicated by replication factor and constraints,
which should be verified via the range status.

The safest way to apply this fix is the following:

1. Add a new node to the cluster (running this commit) and immediately
decommission it.
2. Run the above command against this fresh node.

This approach has the benefit that the node carrying out the work
is not an essential part of the cluster. This means that unexpected
crashes or corruption would likely be confined to this one node.

If this is deemed too much work, an existing node can be used.

In either case, it is indisposable to have taken a recent backup and
to be ready to respond to unanticipated consequences (for example
by restoring the backup).

Release note: None

bdarnell reviewed (approved):
Rubberstamping; some minor comments in the master branch PR.

We should probably backport this to 20.1 too for consistency although there's no definite need for it at the moment. 

> Reviewed 17 of 17 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50387#-:-MA7QhSyE11ioYI0ycS5:b-7ewps)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dbist)



<!-- Sent from Reviewable.io -->

dbist commented:
@bobvawter and I validated the patch on roachprod. The difference here is the tag is targeting 19.2.7 and the cluster in question is 19.2.2, not sure it makes a lot of difference. The other point to note, we picked an arbitrary range to repair, not necessarily a missing range. Any other points we need to address as part of the test?  

# start new node with patched binary
```
./cockroach start --insecure --background --join=10.142.0.29:26257,10.142.0.30:26257,10.142.0.28:26257
```

# immediately decommission, node ID is 7
```
./cockroach node decommission 7 --insecure
```

there are 0 underreplicated ranges

# look up a range
```
show ranges from database tpcc;
```

# we will use an arbitrary range 33
```
./cockroach debug unsafe-heal-missing-range 33 --insecure --host=artem-premier-new-0001
```

The range is on node 7
http://artem-premier-0001.roachprod.crdb.io:26258/#/reports/range/33

# drain this node
```
cockroach node drain --insecure --host=artem-premier-new-0001
```

# decommission the node
```
./cockroach node decommission 7 --insecure
```

# it will run for a while, we're going to ctrl-C out of it
# on the 7th node
```
pkill cockroach
```

```
ubuntu@artem-premier-new-0001:~$ initiating graceful shutdown of server
server drained and shutdown completed
```

after 5minutes, the under replicated range will clear up, the node will move to decom nodes section in the admin ui
```

bdarnell commented:
> The other point to note, we picked an arbitrary range to repair, not necessarily a missing range. 

I'm surprised that using this on a non-missing range didn't make something blow up. 

> Any other points we need to address as part of the test?

We should be sure to test a secure cluster, and to do a query that touches the "healed" range and make sure it's empty.

dbist commented:
> > The other point to note, we picked an arbitrary range to repair, not necessarily a missing range.
> 
> I'm surprised that using this on a non-missing range didn't make something blow up.
> 
> > Any other points we need to address as part of the test?
> 
> We should be sure to test a secure cluster, and to do a query that touches the "healed" range and make sure it's empty.

@bobvawter and I stood up a geo cluster as advised, pinned a table to a region, took down all of the nodes in that region, brought up a new node in that region with the patched binary, ran the repair and were able to query that same table with empty result set. All was done in a secure cluster.

Before the scenario

```
SHOW RANGES FROM TABLE stock;
```

```
  start_key | end_key | range_id | range_size_mb | lease_holder |               lease_holder_locality               | replicas |                                                                      replica_localities
+-----------+---------+----------+---------------+--------------+---------------------------------------------------+----------+---------------------------------------------------------------------------------------------------------------------------------------------------------------+
  NULL      | NULL    |       32 |     38.515228 |            9 | cloud=gce,region=europe-west2,zone=europe-west2-b | {7,8,9}  | {"cloud=gce,region=europe-west2,zone=europe-west2-b","cloud=gce,region=europe-west2,zone=europe-west2-b","cloud=gce,region=europe-west2,zone=europe-west2-b"}
(1 row)

Time: 440.690016ms
```

After repair

```
root@localhost:26257/defaultdb> show ranges from table tpcc.stock;
  start_key | end_key | range_id | range_size_mb | lease_holder |               lease_holder_locality               | replicas |                  replica_localities
+-----------+---------+----------+---------------+--------------+---------------------------------------------------+----------+-------------------------------------------------------+
  NULL      | NULL    |       32 |             0 |           13 | cloud=gce,region=europe-west2,zone=europe-west2-b | {13}     | {"cloud=gce,region=europe-west2,zone=europe-west2-b"}
(1 row)

Time: 143.864517ms
```

dbist reviewed (approved):
the advised scenario worked, the big question is whether we want to proceed with a net new node with patched binary to repair or swap out an existing node.

knz commented:
is this PR still relevant?

knz commented:
is this PR still relevant?

bdarnell commented:
I don't think we want to merge this PR to 19.2 any more, but the last comments on the main branch PR #50268 suggest that there are some updates in this branch that need to be forward-ported to the main branch version (unless we're giving up on getting this merged).

knz commented:
@tbg  could you review this last comment by Ben and indicate which work is remaining? I'm trying to scope out the items that fell into the gaps during roadmapping.

irfansharif commented on pkg/server/node.go:
```diff
@@ -1044,3 +1045,81 @@ func (n *Node) RangeFeed(
+
+	var expValue *roachpb.Value
+	var desc roachpb.RangeDescriptor
+	if err := n.storeCfg.DB.Txn(ctx, func(ctx context.Context, txn *client.Txn) error {
```
The basic primitive should take the range descriptor. We can leave it to the higher level abstraction to give us an appropriate one. There will be two ways to get this range descriptor: either look through meta2 (like we're doing here), or broadcast something cluster-wide and see if there are any surviving replicas for the given range.

irfansharif commented on pkg/server/node.go:
```diff
@@ -1044,3 +1045,81 @@ func (n *Node) RangeFeed(
+	}); err != nil {
+		return nil, err
+	}
+	storeID, nodeID := s.Ident.StoreID, s.Ident.NodeID
```
The basic primitive should also take in the target node+store ID (and leave it to the caller to provide it for us).

irfansharif commented on pkg/server/node.go:
```diff
@@ -1044,3 +1045,81 @@ func (n *Node) RangeFeed(
+	}
+	storeID, nodeID := s.Ident.StoreID, s.Ident.NodeID
+
+	dead := append([]roachpb.ReplicaDescriptor(nil), desc.Replicas().All()...)
```
Note to self: Here we're just constructing the range descriptor with just the one voter, and removing all else.

irfansharif commented on pkg/server/node.go:
```diff
@@ -1044,3 +1045,81 @@ func (n *Node) RangeFeed(
+		desc.RemoveReplica(rd.NodeID, rd.StoreID)
+	}
+
+	if err := n.storeCfg.DB.CPut(
```
Here we're updating the meta2 range with this new dummy descriptor. Is it safe to do this before creating the replica? I think it's fine, we're no more "unsafe" than before.

The basic primitive should also not really touch meta2. We should again rely on higher level abstractions to do this for us.

Note to self: There are two copies of the range descriptor: one in meta2, and one on the range itself. They're usually transactionally consistent. There are some invariants here around how intents are stored, how we recover from finding intents on stores, and whether to assume they're committed or not. I'm a bit fuzzy here.

irfansharif replied:
It's also fine to not update the meta2. The real source of truth is always the range local one. As written we're CPut-ing on the meta2 key. We _should_ be able to CPut on the range local one and then eventually blindly write to the meta2 one. We already rely on something similar for our offline recovery tool (where we CPut to the range local one only).

The danger is that if we don't update meta2, there's a race with the replicaGCqueue. This dummy replica may very well GC itself (after consulting the meta2 range). 

Squinting here a bit further: could a leaseholder read on the local copy be used instead of meta2? It has the nice property of avoiding putting meta2 on the hotpath for online recovery of multiple ranges.

irfansharif reviewed (commented):
X-Ref: https://github.com/cockroachdb/cockroach/pull/50268 for the PR that recovers a range when there _are_ other replicas floating about. Notes below from live review. +cc @TheSamHuang.
